File: c++/T654.cpp
//
// Created by jesse on 2021/1/3.
//
#include <vector>

using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:

    TreeNode *build(vector<int>& nums,int lo,int hi){
        if (lo > hi) {
            return nullptr;
        }
        int max = nums[lo];
        int index=lo;
        for (int i = lo; i <= hi; ++i) {
            if (nums[i] > max) {
                max = nums[i];
                index = i;
            }
        }

        auto l=build(nums,lo,index-1);
        auto r = build(nums, index + 1, hi);
        auto root = new TreeNode(max, l, r);
        return root;
    }
    TreeNode *constructMaximumBinaryTree(vector<int> &nums) {
//        if (nums.empty()) {
//            return NULL;
//        }
//        auto max = std::max_element(nums.begin(), nums.end());
//        vector<int> left(nums.begin(), max);
//        vector<int> right(max + 1, nums.end());
//        auto l=constructMaximumBinaryTree(left);
//        auto r=constructMaximumBinaryTree(right);
//        auto root=new TreeNode(*max,l,r);
//        return root;
        return build(nums,0,nums.size()-1);
    }
};

File: rust/src/t654.rs
/**
给定一个不含重复元素的整数数组。一个以此数组构建的最大二叉树定义如下：

二叉树的根是数组中的最大元素。
左子树是通过数组中最大值左边部分构造出的最大二叉树。
右子树是通过数组中最大值右边部分构造出的最大二叉树。
通过给定的数组构建最大二叉树，并且输出这个树的根节点。

 

示例 ：

输入：[3,2,1,6,0,5]
输出：返回下面这棵树的根节点：

      6
    /   \
   3     5
    \    /
     2  0
       \
        1

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/maximum-binary-tree
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
*/
use std::cell::RefCell;
use std::rc::Rc;
use std::process::exit;
use std::panic::resume_unwind;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution;

impl Solution {
    pub fn construct_maximum_binary_tree(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        Self::build(&nums, 0, nums.len() as i32 - 1)
    }

    pub fn build(nums: &Vec<i32>, start: i32, end: i32) -> Option<Rc<RefCell<TreeNode>>> {
        if start > end {
            return None;

        }
        let mut i = start;
        for j in start..=end {
            if nums[j as usize] > nums[i as usize] {
                i = j;
            }
        }
        Some(Rc::new(RefCell::new(TreeNode {
            val: nums[i as usize],
            left: Self::build(nums, start, i - 1),
            right: Self::build(nums, i + 1, end),
        })))
    }
}